cost-updater/src/main/scala/org/broadinstitute/dsp/workbench/hamm/costUpdater/Main.scala

```diff
@@ -1,62 +1,49 @@
-//        .addService(service)
-//        .addService(ProtoReflectionService.newInstance())
-//        .stream[IO]
-//        .evalMap(server => IO(server.start()))
-//      _ <- Stream(subscribeStream, grpcStream, processMetadataStream).parJoin(50) //TODO: think more about max open
-//    } yield ()
-//
-//    app.handleErrorWith(error => Stream.eval(logger.error(error)("Failed to start server")))
-//      .evalMap(_ => IO.never)
-//      .compile
-//      .drain
-//      .as(ExitCode.Success)
-//  }
-//}
-//
-//class CostUpdaterGrpcImp[F[_]: Sync: Logger] extends CostUpdaterFs2Grpc[F] {
-//  override def status(request: CostUpdaterStatusRequest, clientHeaders: Metadata): F[CostUpdaterStatusResponse] = Sync[F].point(CostUpdaterStatusResponse(
-//    BuildInfo.scalaVersion,
-//    BuildInfo.sbtVersion,
-//    BuildInfo.gitHeadCommit.getOrElse("No commit yet"),
-//    BuildInfo.buildTime,
-//    BuildInfo.toString
-//  ))
-//}
\ No newline at end of file
+package org.broadinstitute.dsp.workbench.hamm
+package costUpdater
+
+import cats.effect._
+import cats.implicits._
+import io.chrisdavenport.log4cats.slf4j.Slf4jLogger
+import fs2._
+import fs2.concurrent.InspectableQueue
+import org.broadinstitute.dsde.workbench.google2.{Event, GoogleStorageService, GoogleSubscriber}
+import org.broadinstitute.dsde.workbench.util.ExecutionContexts
+import MessageProcessor.notificationMessageDecoder
+import org.http4s.implicits._
+import org.http4s.server.Router
+import org.http4s.server.blaze.BlazeServerBuilder
+
+object Main extends IOApp {
+  override def run(args: List[String]): IO[ExitCode] =  {
+    implicit val logger = Slf4jLogger.getLogger[IO]
+
+    val app: Stream[IO, Unit] = for {
+      appConfig <- Stream.fromEither[IO](Config.appConfig)
+
+      _ <- Stream.eval(logger.info("Starting Hamm Cost Updater server"))
+
+      credential <- Stream.resource(org.broadinstitute.dsde.workbench.google2.credentialResource[IO](appConfig.google.subscriber.pathToCredentialJson))
+      blockingExecutionContext <- Stream.resource(ExecutionContexts.fixedThreadPool[IO](256)) //scala.concurrent.blocking has default max extra thread number 256, so use this number to start with
+      storage <- Stream.resource(GoogleStorageService.resource(appConfig.google.subscriber.pathToCredentialJson, blockingExecutionContext))
+
+      queue <- Stream.eval(InspectableQueue.bounded[IO, Event[NotificationMessage]](10))
+      subscriber <- Stream.resource(GoogleSubscriber.resource(appConfig.google.subscriber, queue))
+
+      messageProcessor = MessageProcessor[IO](subscriber, storage)
+
+      service = CostUpdaterService(queue).service
+      serverStream = BlazeServerBuilder[IO]
+        .bindHttp(8080, "0.0.0.0")
+        .withHttpApp(Router("/" -> service).orNotFound)
+        .serve
+        .void
+       _ <- Stream(Stream.eval(subscriber.start), messageProcessor.process, serverStream).parJoin(10) //TODO: potentially adjust maxOpen
```

Comment:
I need stream of stream because I want to run the subscriber stream, server stream, and the stream for processing metadata at the same time. You're right about parJoin, it should be 3 not 10.